proxy/http/HttpSM.cc

```diff
@@ -7988,3 +7987,58 @@ HttpSM::find_proto_string(HTTPVersion version) const
   }
   return nullptr;
 }
+
+// YTS Team, yamsat Plugin
+// Function to copy the partial Post data while tunnelling
+void
+PostDataBuffers::copy_partial_post_data()
+{
+  this->postdata_copy_buffer->write(this->ua_buffer_reader);
+  Debug("http_redirect", "[PostDataBuffers::copy_partial_post_data] wrote %" PRId64 " bytes to buffers %" PRId64 "",
+        this->ua_buffer_reader->read_avail(), this->postdata_copy_buffer_start->read_avail());
+  this->ua_buffer_reader->consume(this->ua_buffer_reader->read_avail());
+}
+
+// YTS Team, yamsat Plugin
+// Allocating the post data buffers
+void
+PostDataBuffers::init(IOBufferReader *ua_reader)
+{
+  // the this->ua_buffer_reader may be realeased in tunnel::deallocate_buffers.
+  // we always set a new reader instead of the old value;
+  // ink_assert(this->ua_buffer_reader == nullptr);
+
+  Debug("http_redirect", "[PostDataBuffers::init]");
+
+  this->ua_buffer_reader = ua_reader;
+
+  if (this->postdata_copy_buffer == nullptr) {
+    ink_assert(this->postdata_copy_buffer_start == nullptr);
+    this->postdata_copy_buffer       = new_empty_MIOBuffer(BUFFER_SIZE_INDEX_4K);
+    this->postdata_copy_buffer_start = this->postdata_copy_buffer->alloc_reader();
+  }
+}
+
+// YTS Team, yamsat Plugin
+// Deallocating the post data buffers
+void
+PostDataBuffers::clear()
+{
+  Debug("http_redirect", "[PostDataBuffers::clear]");
+
+  if (this->postdata_copy_buffer != nullptr) {
+    free_MIOBuffer(this->postdata_copy_buffer);
+    this->postdata_copy_buffer       = nullptr;
+    this->postdata_copy_buffer_start = nullptr; // deallocated by the buffer
+  }
+
+  if (this->ua_buffer_reader != nullptr) {
+    this->ua_buffer_reader->mbuf->dealloc_reader(this->ua_buffer_reader);
```

Comment:
Do we need to check that `this->ua_buffer_reader->mbuf != nullptr`?

Reply:
I think `this->ua_buffer_reader->mbuf` should not be nullptr !  
I'll add a assertion here !

Reply:
Yes, the `ua_buffer_reader` may be released in `tunnel`.